src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetInfoService.cs

```diff
@@ -111,52 +112,95 @@ public virtual bool ShouldFormatSnippet(SnippetInfo snippetInfo)
 
         private void PopulateSnippetCaches()
         {
-            var updatedSnippets = GetSnippetInfoList();
-            var updatedSnippetShortcuts = GetShortcutsHashFromSnippets(updatedSnippets);
+            Debug.Assert(_expansionManager != null);
 
-            lock (cacheGuard)
-            {
-                snippets = updatedSnippets;
-                snippetShortcuts = updatedSnippetShortcuts;
-            }
-        }
+            // Ideally we'd fork execution here based on whether the expansion manager is an
+            // IExpansionManager or not. Unfortunately, we cannot mention that type by name until
+            // the Roslyn build machines are upgraded to Visual Studio 2015 Update 1. We therefore
+            // need to try using IExpansionManager dynamically, from a background thread. If that
+            // fails, then we come back to the UI thread for using IVsExpansionManager instead.
 
-        protected static HashSet<string> GetShortcutsHashFromSnippets(IList<SnippetInfo> updatedSnippets)
-        {
-            return new HashSet<string>(updatedSnippets.Select(s => s.Shortcut), StringComparer.OrdinalIgnoreCase);
+            var token = _waiter.BeginAsyncOperation(GetType().Name + ".Start");
+
+            Task.Factory.StartNew(async () => await PopulateSnippetCacheOnBackgroundWithForegroundFallback().ConfigureAwait(false),
+                            CancellationToken.None,
+                            TaskCreationOptions.None,
+                            TaskScheduler.Default).CompletesAsyncOperation(token);
         }
 
-        private IList<SnippetInfo> GetSnippetInfoList()
+        private async Task PopulateSnippetCacheOnBackgroundWithForegroundFallback()
         {
-            IVsExpansionEnumeration expansionEnumerator;
-            if (TryGetVsSnippets(out expansionEnumerator))
+            AssertIsBackground();
+
+            try
+            {
+                IVsExpansionEnumeration expansionEnumerator = await ((dynamic)_expansionManager).EnumerateExpansionsAsync(
+                    _languageGuidForSnippets,
+                    0, // shortCutOnly
+                    Array.Empty<string>(), // types
+                    0, // countTypes
+                    1, // includeNULLTypes
+                    1 // includeDulicates: Allows snippets with the same title but different shortcuts
+                    ).ConfigureAwait(false);
+
+                // The rest of the process requires being on the UI thread
```

Comment:
I'm assuming we don't want to rely on COM marshalling behavior either because of risk of deadlock, or just because the interface is really chatty?

I'm not saying this is unreasonable (it actually seems quite reasonable), I just want the comment here to be updated to capture the "why" instead of the "what". The code is clear it's going to the UI thread without the comment, but just want to capture the reason.